pkg/gadget-service/instance-manager/file-store/filestore.go

```diff
@@ -0,0 +1,185 @@
+	}
+	return res, nil
+}
+
+func (s *FileStore) LoadStoredGadgets() error {
+	gadgets, err := s.getGadgets()
+	if err != nil {
+		return fmt.Errorf("reading existing gadgets: %w", err)
+	}
+
+	for _, gadget := range gadgets {
+		log.Infof("loading gadget instance %q", gadget.GadgetInstance.Id)
+		s.instanceMgr.RunGadget(gadget.GadgetInstance)
+	}
+	return nil
+}
+
+func (s *FileStore) InstallGadgetInstance(ctx context.Context, req *api.InstallGadgetInstanceRequest) (*api.InstallGadgetInstanceResponse, error) {
+	idBytes := make([]byte, 16)
+	_, err := io.ReadFull(rand.Reader, idBytes)
+	if err != nil {
+		return nil, errors.New("could not create gadget id")
+	}
+	id := hex.EncodeToString(idBytes)
+	req.GadgetInstance.Id = id
+
+	if req.GadgetInstance.Name == "" {
+		req.GadgetInstance.Name = namesgenerator.GetRandomName(0)
+	}
+
+	// Store to gadget file
+	gadgetBlob, _ := protojson.Marshal(req)
+	filename := filepath.Join(GadgetFilePath, fmt.Sprintf("%s.gadget", id))
+	err = os.WriteFile(filename, gadgetBlob, 0o644)
+	if err != nil {
+		return nil, fmt.Errorf("storing gadget information: %w", err)
+	}
+
+	log.Debugf("installing new gadget %q", id)
+	s.instanceMgr.RunGadget(req.GadgetInstance)
+	return &api.InstallGadgetInstanceResponse{
+		Result:         0,
+		GadgetInstance: req.GadgetInstance,
+	}, nil
+}
+
+func (s *FileStore) ListGadgetInstances(ctx context.Context, request *api.ListGadgetInstancesRequest) (*api.ListGadgetInstanceResponse, error) {
+	gadgets, err := s.getGadgets()
+	if err != nil {
+		return nil, fmt.Errorf("loading gadgets: %w", err)
+	}
+	persistentGadgets := make([]*api.GadgetInstance, 0, len(gadgets))
+	for _, gadget := range gadgets {
+		persistentGadgets = append(persistentGadgets, gadget.GadgetInstance)
+	}
+	return &api.ListGadgetInstanceResponse{GadgetInstances: persistentGadgets}, nil
+}
+
+func (s *FileStore) GetGadgetInstance(ctx context.Context, req *api.GadgetInstanceId) (*api.GadgetInstance, error) {
+	path := filepath.Join(GadgetFilePath, fmt.Sprintf("%s.gadget", req.Id))
+	gadget, err := loadGadgetFile(path)
+	if err != nil {
+		return nil, fmt.Errorf("loading gadget: %w", err)
+	}
+	return gadget.GadgetInstance, nil
+}
+
+func (s *FileStore) RemoveGadgetInstance(ctx context.Context, request *api.GadgetInstanceId) (*api.StatusResponse, error) {
+	path := filepath.Join(GadgetFilePath, fmt.Sprintf("%s.gadget", request.Id))
+	_, err := loadGadgetFile(path)
+	if err != nil {
+		return &api.StatusResponse{Result: 1, Message: err.Error()}, nil
+	}
+
+	err = s.instanceMgr.RemoveGadget(request.Id)
+	if err != nil {
+		return &api.StatusResponse{Result: 1, Message: err.Error()}, nil
+	}
+	err = os.Remove(path)
+	if err != nil {
+		return &api.StatusResponse{Result: 1, Message: err.Error()}, nil
+	}
+	return &api.StatusResponse{Result: 0}, nil
+}
+
+func (s *FileStore) StopGadgetInstance(ctx context.Context, request *api.GadgetInstanceId) (*api.StatusResponse, error) {
+	return &api.StatusResponse{Result: 0}, nil
```

Comment:
Make this return an implemented error instead until it's implemented?

Reply:
I removed the Stop() feature for now as it is not completely clear from the UX how it should behave.